playbooks/roles/mongo_3_2/templates/backup-mongo.sh.j2

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+# ref https://tasks.opencraft.com/browse/SE-1669
+# Script to perform a point-in-time dump of the local mongodb database using
+# mongodump.
+# includes locking (prevent this script from running multiple times in
+# parallel), creating a snapshot of the volume used to backup to, and snitch monitoring
+
+# exit by default on failure
+set -e
+
+# vars
+ARCHIVE_NAME="mongo-backup-$(date --iso-8601=minutes --utc).gz"
+VOLUME="{{ MONGO_BACKUP_EBS_VOLUME_DEVICE }}" # eg. /dev/svdk
+VOLUME_ID="{{ MONGO_BACKUP_EBS_VOLUME_ID }}" # eg. vol-123456
+VOLUME_MOUNT_PATH="{{ MONGO_BACKUP_VOLUME_MOUNT_PATH }}" # eg. vol-123456
+SNITCH_URL="{{ MONGO_BACKUP_SNITCH_URL }}"
+MONGO_USER="{{ COMMON_MONGO_READ_ONLY_USER }}"
+MONGO_PASS="{{ COMMON_MONGO_READ_ONLY_PASS }}"
+HOSTNAME="{{ MONGO_BACKUP_HOST }}"
+
+# TODO: is hostname what we want here?
+if [ "$HOSTNAME" = $(hostname) ]; then
+    echo "This is not the backup host. Run on a different instance."
+    exit 1
+fi
+
+# Acquire backup lock using this script itself as the lockfile. If another
+# backup task is already running, then exit immediately.
+exec 200<$0
+flock -n 200 || { echo "Another backup task is already running."; exit 1; }
+
+echo "Starting at $(date)"
+
+# ensure volume is mounted
+mkdir -p "$VOLUME_MOUNT_PATH"
+if ! mountpoint -q "$VOLUME_MOUNT_PATH"; then
+  mount -o discard,defaults,noatime "$MONGO_BACKUP_EBS_VOLUME_DEVICE" "$VOLUME_MOUNT_PATH"
+fi
+
+# clean old backup files to save space and so we start afresh
+rm -rf "$VOLUME_MOUNT_PATH/mongo/"
+
+
+# create the dump
+# XXX: we may want to check how this lays out of disk and how it will play against ebs volume snapshots. The idea was
+# that snapshots would be cheap because you only pay for data blocks that have changed between one snapshot and the
+# next. however, if the mongodump -> gzip -> put in a single archive process ends up not being consistent in layout,
+# this may end up with a lot of the disk content changing between snapshots.
+# TODO: investigate which would be better: a single archive (as below) or a directory with individual files. It's
+# possible that a directory with individual files may be better for performance.
+mongodump --host="localhost" --opLog --gzip -u "$MONGO_USER" --password "$MONGO_PASS" --authenticationDatabase "admin" --archive="$VOLUME_MOUNT_PATH/mongo/$ARCHIVE_NAME"
+
+
+# flush everything to disk, and unmount the volume ready to snapshot
+sync
+umount "$VOLUME_MOUNT_PATH"
+
+# create a snapshot of the volume
+aws ec2 create-snapshot --volume-id "$VOLUME_ID"
+# XXX: this command returns immediately, so we have a period of time between now and when the snapshot is finished
```

Comment:
@pomegranited would this be a problem? How long do ebs snapshots take to create?

Reply:
Snapshot creation time depends on the size of the volume.. so this could be a while.

Reply:
This is a bit boring, but I think we're going to have to poll until the snapshot is complete: [describe-snapshots](https://docs.aws.amazon.com/cli/latest/reference/ec2/describe-snapshots.html#examples)

Reply:
No problem. :)

By the way, can we add installing [jq](https://stedolan.github.io/jq/) to the instances? It's so much nicer doing something like `cat | jq -r .VolumeId` rather than awk/sed/cut/etc. to work with the json output of awscli.

Reply:
Sure, that seems reasonable.